src/log.cpp

```diff
@@ -0,0 +1,182 @@
+#include "log.h"
+
+namespace bpftrace {
+
+std::string logtype_str(LogType t)
+{
+  switch (t)
+  {
+    // clang-format off
+    case LogType::debug   : return "DEBUG";
+    case LogType::info    : return "INFO";
+    case LogType::warning : return "WARNING";
+    case LogType::error   : return "ERROR";
+    // clang-format on
+    default:
+      std::cerr << "Log type not found." << std::endl;
+      abort();
+  }
+}
+
+Log::Log()
+{
+}
+
+Log::~Log()
+{
+}
+
+Log& Log::get()
+{
+  static Log log;
+  return log;
+}
+
+void Log::take_input(LogType type, std::string&& input, std::ostream& out)
+{
+  out << logtype_str(type) << ": " << input << std::endl;
+}
+
+void Log::take_input(LogType type,
+                     std::string&& input,
+                     const location& loc,
+                     std::ostream& out)
+{
+  if (src_.empty())
+  {
+    std::cerr << "Cannot use LLOG before setting source for Log" << std::endl;
+    abort();
+  }
+  log_with_location(type, loc, out, input);
+}
+
+const std::string Log::get_source_line(unsigned int n)
+{
+  // Get the Nth source line (N is 0-based). Return an empty string if it
+  // doesn't exist
+  std::string line;
+  std::stringstream ss(src_);
+  for (unsigned int idx = 0; idx <= n; idx++)
+  {
+    std::getline(ss, line);
+    if (ss.eof() && idx == n)
+      return line;
+    if (!ss)
+      return "";
+  }
+  return line;
+}
+
+void Log::log_with_location(LogType type,
+                            const location& l,
+                            std::ostream& out,
+                            const std::string& m)
+{
+  if (filename_ != "")
```

Comment:
Style nit: "positive" conditions tend to be easier to read
```suggestion
  if (filename_.size())
```